feat: floor price set to the 0.06 USDFC/month

rvagg opened the pull request:
_Edit: This got switched to having 0.06 USDFC as the floor, not the bytes; so now the bytes works out as ~24.576 GiB and we get to talk about it in terms of cents per month._

(PR on top of #316, that should be merged first)

Amounts below this pay as if they are storing 24 GiB, which is 0.05859375 USDFC / month.

Closes: https://github.com/FilOzone/filecoin-services/issues/319

---

*  24 GiB = 24/1024 TiB = 0.0234375 TiB
* At 2.5 USDFC/TiB/month, price = 0.0234375 * 2.5 = 0.05859375 USDFC/month

Are we OK with 0.05859375? I have an option here of either setting 24GiB as the floor _size_ or 0.06 as the floor _price_.

If we go with 0.06 as the floor then we're working with 24.576 GiB floor size.

---

I also had to change the mock token from 6 decimal places to 18, cause there's already a floor rounding bit of code in here and it was getting stuck on a higher floor due to not enough decimal places, so everything came out wrong.

rvagg committed:
feat: floor price set to 0.06 USDFC/month per data set

For storage amounts below ~24.576 GiB, the rate will be 0.06 USDFC / month.
Above this amount the price is per-byte.

Closes: https://github.com/FilOzone/filecoin-services/issues/319

Kubuxu commented:
Is it feasible to add a check that the floor price rate is available to be paid when doing `createDataSet`? It would solve one of the problems causing: https://github.com/filecoin-project/curio/issues/726

wjmelements commented on service_contracts/src/FilecoinWarmStorageService.sol:
```diff
@@ -1200,18 +1203,26 @@ contract FilecoinWarmStorageService is
+        uint256 naturalRate = calculateStorageSizeBasedRatePerEpoch(totalBytes, STORAGE_PRICE_PER_TIB_PER_MONTH);
+
+        // Calculate minimum rate (floor price converted to per-epoch)
+        uint256 minimumRate = MINIMUM_STORAGE_RATE_PER_MONTH / EPOCHS_PER_MONTH;
```
this might be zero. how many decimals are there? If only 6, this is zero.

wjmelements replied:
I have checked; there are 18, so we are good
https://filfox.info/en/token/f410fqc4y2ovat777evodxjfcieir74jgf4cfcyah4wq?t=contract

rvagg commented on service_contracts/src/FilecoinWarmStorageService.sol:
```diff
@@ -1200,18 +1203,26 @@ contract FilecoinWarmStorageService is
+        uint256 naturalRate = calculateStorageSizeBasedRatePerEpoch(totalBytes, STORAGE_PRICE_PER_TIB_PER_MONTH);
+
+        // Calculate minimum rate (floor price converted to per-epoch)
+        uint256 minimumRate = MINIMUM_STORAGE_RATE_PER_MONTH / EPOCHS_PER_MONTH;
```
oh, thanks for the reminder, it's not supposed to be the only place, we need to expose it via `getServicePrice()` for the SDK, and we need it per month there and we can't multiply it out to reconstitute it because there's rounding problems with per-epoch (that we'll mostly ignore for now), so we should keep storing MINIMUM_STORAGE_RATE_PER_MONTH and I'll just add that accessor to ServicePricing

Kubuxu replied:
It might get optimised out

Kubuxu replied:
I'm almost certain you are right. Still, I'm ok paying that cost of division right now.

rvagg committed:
feat: check payer has minimum available balance before creating data set

rvagg commented:
> Is it feasible to add a check that the floor price rate is available to be paid when doing createDataSet?

Oh, this is a really good idea, and yes it is possible! We have a `getAccountInfoIfSettled` on FilecoinPay that we can use to check their balance (if they were fully settled) so we can compare that with what would be needed for initial lockup and reject early. With create+add this is nice and clean. With create, then add, it's a little more messy and we lose guarantees but you'd just get rejected at add if you used up funds in between the two calls. If nextProvingPeriod gets done immediately after create+add then Curio should have a pretty solid guarantee that there is at least 0.06 on the table if they prove for a month, with a slight window for a possible race.

See latest commit in here for that feature.

Kubuxu commented:
Does that also cover `OperatorRateAllowanceExceeded`?

rvagg committed:
feat(warm-storage): validate operator approvals in dataSetCreated

Add comprehensive validation of operator approval settings alongside
existing minimum balance check during createDataSet operation.
Validates operator is approved with sufficient rate allowance, lockup allowance,
and max lockup period.

rvagg committed:
fix: s/calculateRatesPerEpoch/calculateRatePerEpoch

rvagg commented:
Figured out how to add additional checks in there. I'm of two minds about this, mostly it's great UX, we catch errors early and can be descriptive about it - good for SPs and good for users. But it also adds complexity and gets FWSS more deeply involved in something that it feels like should be something should be able to ask Filecoin Pay to figure out for us. But it's probably fine.

* `InsufficientFundsForMinimumRate` - not enough deposit
* `OperatorNotApproved` - need to approve FWSS
* `InsufficientRateAllowance` - not enough approval for the minimum rate (0.06/month)
* `InsufficientLockupAllowance` - not enough lockup approval for minimum rate @ 30 days
* `InsufficientMaxLockupPeriod` - not enough lockup period approval for the 30 day lockup

Kubuxu commented:
> But it's probably fine.

We will be able to remove it when we combine `createDataSet` and initial `nextProvePeriod`.

Kubuxu reviewed (approved):
SGTM

wjmelements commented on service_contracts/src/FilecoinWarmStorageService.sol:
```diff
@@ -1177,21 +1234,29 @@ contract FilecoinWarmStorageService is
      */
-    function calculateRatesPerEpoch(uint256 totalBytes) external view returns (uint256 storageRate) {
-        storageRate = calculateStorageSizeBasedRatePerEpoch(totalBytes, STORAGE_PRICE_PER_TIB_PER_MONTH);
+    function calculateRatePerEpoch(uint256 totalBytes) external view returns (uint256 storageRate) {
```
I think this method is unused. I plan to move it to View as part of the #469 refactors

wjmelements replied:
https://github.com/FilOzone/filecoin-services/pull/476
